lib/libzpool/uzfs_io.c

```diff
@@ -21,7 +21,43 @@
 
 #include <sys/dmu_objset.h>
 #include <sys/uzfs_zvol.h>
+#include <sys/zil_impl.h>
 
+extern ssize_t zvol_immediate_write_sz;
+
+/*
+ * checks for IO overlap with any ongoing sync IOs
+ * lock handling need to be done by caller
+ */
+static boolean_t
+check_io_overlap_with_sync_list(zvol_state_t *zv, uint64_t r_offset,
+    uint64_t r_len)
+{
+	dmu_sync_node_t  *syncnode, *p_syncnode;
+
+	syncnode = list_head(&zv->zv_dmu_sync_list);
+	while (syncnode != NULL) {
+		add_ref_cnt(syncnode);
+		if (syncnode->offset < r_offset) {
+			if (syncnode->end > r_offset) {
+				drop_ref_cnt(syncnode);
+				return (B_TRUE);
+			}
+		} else if (syncnode->offset == r_offset) {
+			drop_ref_cnt(syncnode);
+			return (B_TRUE);
+		} else {
+			if ((r_offset + r_len) > syncnode->offset) {
```

Comment:
if (syncnode->offset, syncnode->end, r_offset, r_end) then also this condition holds true.
Should we allow this case as it is not overlapping?

Reply:
this case is handled in first 'if' itself